### DISCUSSION TITLE
ダイナミックインポート`import() `とは？

### DISCUSSION THREAD

**Discussion by kazzyfrog** [lgtm-factory/lgtm-factory#83] [Scrap]
全員に共有しておいた方が良さそうなこと & 自分もまだ理解が浅いこと
なので、つらつら記録していきますー👍

参考
- https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Operators/import
- https://qiita.com/tonkotsuboy_com/items/f672de5fdd402be6f065
- https://zenn.dev/ignorant_kenji/articles/f60ef36bf84077

**Comment by kagomen** [comment]
参考記事より
https://zenn.dev/ignorant_kenji/articles/f60ef36bf84077

> Dynamic Import は必要な時に、必要なファイルを読み込みます。
Dynamic Importの使用を検討する必要がある検討箇所を複数述べます。
static importすると、コードの読み込みが大幅に遅くなり、インポートするコードが必要になる可能性が低くなるか、後で必要になる可能性が低くなります。
static importすると、プログラムのメモリー使用量が大幅に増加し、インポートするコードが必要になる可能性が低くなります。
読み込み時に読み込むファイルが存在しないケース
読み込むファイルに副作用があり、ある条件がtrueではない限り、その副作用を望まない場合。（ファイルに副作用がないことを推奨しますが、ファイルの依存関係でこれを制御できない場合があります。）
すべての場合において、Dynamic import を使用するのは賢い判断ではなく、Dynamic import は必要な場合にのみ使用してください。
初期の依存関係をロードするにはstatic importが適していて、tree shakingの恩恵をよりよく受けることができます。

**Comment by kagomen** [comment]
## reactでdynamic importを使うときは、lazy()を使わないといけないのですか？

必ずしも`lazy()`を使う必要はありませんが、`lazy()`を使用することで多くの利点があります。

1. `lazy()`を使わない場合:

Dynamic importを直接使用することも可能です。例えば：

```javascript
import('./SomeComponent').then(module => {
  const Component = module.default;
  // Componentを使用する
});
```

2. `lazy()`を使う場合:

```javascript
const SomeComponent = React.lazy(() => import('./SomeComponent'));
```

`lazy()`を使う利点:

1. コード分割: `lazy()`はReactの組み込み機能で、コード分割を簡単に実現できます。

2. Suspenseとの統合: `lazy()`で読み込んだコンポーネントは`Suspense`と一緒に使用でき、ローディング状態を簡単に管理できます。

   ```jsx
   <Suspense fallback={<div>Loading...</div>}>
     <SomeComponent />
   </Suspense>
   ```

3. 宣言的: コンポーネントの定義がより宣言的になり、読みやすくなります。

4. エラーバウンダリとの連携: `lazy()`で読み込んだコンポーネントはエラーバウンダリでキャッチできるエラーを投げます。

5. TypeScriptサポート: `lazy()`はTypeScriptと相性が良く、型安全性を保ちやすいです。

結論として、`lazy()`を使用することで、Reactのエコシステムとより良く統合され、コードの可読性と保守性が向上します。ただし、特殊なケースや低レベルの制御が必要な場合は、直接dynamic importを使用することも可能です。

**Reply by kagomen** [reply]
https://ja.react.dev/reference/react/lazy#suspense-for-code-splitting

> （lazyを）用いる場合は、遅延インポートしようとしているコンポーネントが default でエクスポートされている必要があります。

`import('./SomeComponent')`だからそうか
名前付きエクスポートはだめだと

**Reply by kagomen** [reply]
https://ja.react.dev/reference/react/lazy#my-lazy-components-state-gets-reset-unexpectedly

> lazy コンポーネントを他のコンポーネントの内部で宣言しないでください。
代わりに、常にモジュールのトップレベルで宣言してください。

🫡

**Reply by kagomen** [reply]
こちらも参考になりそう

今後の React ではどの範囲を Suspense で囲むかという設計が重要になってくる - Zenn
https://zenn.dev/tm35/articles/0a64177c0a41bd

**Comment by kazzyfrog** [comment]
現状、`_plarground`では、
以下のような感じで、雑にダイナミックインポートしてます！

```jsx
  // urlのクエリでテーマ指定
  const searchParams = request.nextUrl.searchParams;
  const type = searchParams.get("type");

  // typeに応じたファイルを、ダイナミックインポート
  const { getImage } = await import(`../../../../jsx-data/${type}`);
  const { jsx, options } = await getImage(emoji);
```

これによって、URLに応じたファイルを、APIリクエストのタイミングで読み込める！

**Reply by kagomen** [reply]
了解ですー！👍🙏

**Comment by kazzyfrog** [comment]
## デフォルトを、ダイナミックインポートする場合
> 返されたオブジェクトから "default" のキーを再構築し、名前を変更する必要があります。

```js
(async () => {
  if (somethingIsTrue) {
    const {
      default: myDefault,
      foo,
      bar,
    } = await import("/modules/my-module.js");
  }
})();
```

知らなかった！

参考：
https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Operators/import#%E3%83%87%E3%83%95%E3%82%A9%E3%83%AB%E3%83%88%E3%82%92%E3%82%A4%E3%83%B3%E3%83%9D%E3%83%BC%E3%83%88
